snuba/query/parser/__init__.py

```diff
@@ -0,0 +1,109 @@
+        # has an issue, anyway the production query is ran based on the old query
+        # representation.
+        # Once we will be actually using the ast to build the Clickhouse query
+        # this try/except block will disappear.
+        enforce_validity = state.get_config("query_parsing_enforce_validity", 0)
+        if enforce_validity:
+            raise e
+        else:
+            logger.error("Failed to parse query", exc_info=e)
+            source = dataset.get_dataset_schemas().get_read_schema().get_data_source()
+            return Query(body, source)
+
+
+def _parse_query_impl(body: MutableMapping[str, Any], dataset: Dataset,) -> Query:
+    aggregate_exprs = []
+    for aggregation in body.get("aggregations", []):
+        assert isinstance(aggregation, (list, tuple))
+        assert len(aggregation) == 3
+        aggregation_function = aggregation[0]
+        assert isinstance(aggregation_function, str)
+        column_expr = aggregation[1]
+        column_expr = column_expr if column_expr else []
+        alias = aggregation[2]
+        assert isinstance(alias, str)
+        alias = alias if alias else None
+
+        aggregate_exprs.append(
+            parse_aggregation(aggregation_function, column_expr, alias)
+        )
+
+    groupby_exprs = [
+        parse_expression(tuplify(group_by)) for group_by in body.get("groupby", [])
+    ]
+    select_exprs = [
+        parse_expression(tuplify(select)) for select in body.get("selected_columns", [])
+    ]
+
+    selected_cols = groupby_exprs + aggregate_exprs + select_exprs
+
+    arrayjoin = body.get("arrayjoin")
+    if arrayjoin:
+        array_join_expr: Optional[Expression] = parse_expression(body["arrayjoin"])
+    else:
+        array_join_expr = None
+
+    where_expr = parse_conditions_to_expr(
+        body.get("conditions", []), dataset, arrayjoin
+    )
+    having_expr = parse_conditions_to_expr(body.get("having", []), dataset, arrayjoin)
+
+    orderby_exprs = []
+    for orderby in body.get("orderby", []):
+        if isinstance(orderby, str):
+            match = NEGATE_RE.match(orderby)
+            assert match is not None, f"Invalid Order By clause {orderby}"
+            direction, col = match.groups()
+            orderby = col
+        elif is_function(orderby):
+            match = NEGATE_RE.match(orderby[0])
+            assert match is not None, f"Invalid Order By clause {orderby}"
+            direction, col = match.groups()
+            orderby = [col] + orderby[1:]
+        else:
+            raise ValueError(f"Invalid Order By clause {orderby}")
+        orderby_parsed = parse_expression(tuplify(orderby))
+        orderby_exprs.append(
+            OrderBy(
+                OrderByDirection.DESC if direction == "-" else OrderByDirection.ASC,
+                orderby_parsed,
+            )
+        )
+
+    source = dataset.get_dataset_schemas().get_read_schema().get_data_source()
+    return Query(
+        body,
```

Comment:
Assuming that the formed queries appear to be correct (aside: how are we planning on validating that?) what still will need to be done before we can stop passing the `body` through here?

Reply:
A few things:
- port all the query processors. Now there is logic done in column_expr that is dataset specific that has to become query processors like tag resolution
- port all query extensions, that now are just updating the body
- address split query, which is tweaking the body 
- now we have a query that means the same thing and could be run against clickhouse
- unfortunately there is no easy way to run both parser and compare the strings. They will be different, as they should be simply because the ast does aliasing properly. So my plan is to roll it out on selected referrers and on selected projects so that we start using it internally. At the same time, I will pick a small sample (a very small one) of production queries and run them in parallel twice against clickhouse and see how different the results are